### PR TITLE
[Teo Sin Yee]  iP

### PR DESCRIPTION
**Pull request by tsinyee**
### **DukePro**

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. Download the `.jar` file from [here](https://github.com/tsinyee/ip/releases/tag/A-Jar).
2. Run it using `java -jar ip.jar`.
3. Add your tasks.
4. Let it manage your tasks for you! 😉

Features:

- [x]  Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

Built on very easy to understand code. Example snippet of the `main` function:

```
    public static void main (String[] args) {
        Duke duke = new Duke();
        duke.start();
    }
```

### CODE REVIEW

File: src/main/java/duke/command/Command.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import duke.util.Storage;
+import duke.util.TaskHandler;
+import duke.util.Ui;
+
+/**
+ * This class encapsulates all commands that Duke can handle.
+ *
+ * @author Teo Sin Yee
+ * @version CS2103T AY21/22 Semester 1
+ */
```
Comment (chowRuiYan) [inline src/main/java/duke/command/Command.java:13]: I like how you separate all the different commands into individual files, making it very neat and easy to read. 👍

File: src/main/java/duke/util/Parser.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+     * Parses the raw user input and returns the corresponding Command to execute.
+     *
+     * @param input Raw user input.
+     * @return The corresponding Command.
+     * @throws DukeException If user input is unexpected.
+     */
+    public Command parseRawInput(String input) throws DukeException {
+        String[] splitInput = input.split("\\s+");
+        String commandWord = splitInput[0];
+        switch (commandWord) {
```
Comment (chowRuiYan) [inline src/main/java/duke/util/Parser.java]: Perhaps you could make the input case-insensitive. 
```suggestion
        switch (commandWord.toLowerCase()) {
```
So instead of throwing a `DukeNoSuchCommandException()` when, for example, *"List"* is given as input, it can still work as expected.

**Review by chowRuiYan — COMMENTED** [review]
I think your code is very well written and I don't have much to comment on. Great job! 😄

File: src/main/java/duke/task/Deadline.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        super(name);
+        this.dueDate = dueDate;
+    }
+
+    public String formatToSave() {
+        return String.format("D | %s | %s", super.formatToSave(), dueDate);
+    }
+
+    /**
+     * String representation of a deadline.
```
Comment (chetwinlow) [inline src/main/java/duke/task/Deadline.java]: Maybe the comment could be changed to "Returns a string ..."?

File: src/main/java/duke/util/Parser.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        String[] eventDetails = input.split("\\s+/at\\s+");
+        if (eventDetails.length < 2 || eventDetails[1].isEmpty()) {
+            throw new DukeEventMissingDateException();
+        } else {
+            return eventDetails;
+        }
+    }
+
+    /**
+     * Extract the date for deadline and event commands.
```
Comment (chetwinlow) [inline src/main/java/duke/util/Parser.java]: Probably careless but you probably meant"Extracts..." ?

File: src/main/java/duke/util/TaskHandler.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        taskList.add(task);
+        return String.format(TASK_ADDED_MESSAGE, task.toString(), getTotalTasksCount());
+    }
+
+    /**
+     * Gets the number of incomplete tasks.
+     *
+     * @return The number of incomplete tasks.
+     */
+    private int getUndoneTasksCount() {
```
Comment (chetwinlow) [inline src/main/java/duke/util/TaskHandler.java]: Might seem a bit wordy? May I suggest something like "getNumUndone"?

File: src/main/java/duke/util/Storage.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                loadedTasks.add(formatToTaskList(sc.nextLine()));
+            }
+            return loadedTasks;
+        } catch (FileNotFoundException e) {
+            newFile();
+            return new ArrayList<>();
+        }
+    }
+
+    /** Creates a new file to store data **/
```
Comment (chetwinlow) [inline src/main/java/duke/util/Storage.java]: I do believe that the comment should end with "*/"?

**Review by chetwinlow — COMMENTED** [review]
Overall code is clean and well formatted!
